fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PullUpJoinFromUnion.java

```diff
@@ -0,0 +1,585 @@
+                for (Slot slot : commonSide.getOutput()) {
+                    commonJoinSlotMap.put(slot, slot);
+                }
+            } else {
+                for (int j = 0; j < commonSide.getOutput().size(); ++j) {
+                    commonJoinSlotMap.put(commonSide.getOutput().get(j), firstJoinOutput.get(j));
+                }
+            }
+        }
+        return commonJoinSlotMap;
+    }
+
+    private class ExpressionOrIndex {
+        Expression exprFromCommonSide = null;
+        int indexOfNewUnionOutput = -1;
+
+        private ExpressionOrIndex(Expression expr) {
+            exprFromCommonSide = expr;
+        }
+
+        private ExpressionOrIndex(int index) {
+            indexOfNewUnionOutput = index;
+        }
+    }
+
+    // In the union child output, the number of outputs from the common side must be the same in each child output,
+    // and the outputs from the common side must be isomorphic (both a+1) and have the same index in the union output.
+    // In the union child output, the number of outputs from the non-common side must also be the same,
+    // but they do not need to be isomorphic.
+    // Output parameters1: otherOutputsList stores the outputs of the other side. The length of each element
+    // in otherOutputsList must be the same.
+    // The i-th element represents the output of the other side in the i-th child of the union.
+    // otherOutputsList is Used to create child nodes of a new Union in the constructNewUnion function.
+    // Output parameter2: upperProjectExpressionOrIndex, used in the constructNewProject function
+    // of creating the top-level project,
+    // records the output column order of the original union, and is used on the basis of the new join output,
+    // setting the column or expression that should be output in the upper-level project operator.
+    // The size of upperProjectExpressionOrIndex must be the same as the output size of the original union.
+    // Pair.first in List represents whether the output comes from the common side or the other side.
+    // True represents the output from the common side, and false represents the output from the other side.
+    // Pair.second in List is a structure. When Pair.first is true,
+    // Pair.second saves the output expression of the common side.
+    // When Pair.first is false, Pair.second saves the output subscript of the other side.
+    // Because the output of the new union has not been constructed at this time, the index is saved.
+    // Since the check part of this function ensures that the outputs at the same position in the union children
+    // must all come from the common side or from the other side,
+    // and when the join is constructed at the end, the common side will use the common side of the first join,
+    // so we only need to fill in upperProjectExpressionOrIndex when processing the output of the first child.
+    private boolean checkUnionChildrenOutput(LogicalUnion union,
```

Comment:
1. please give rejected patterns by example
2. it is better to return output parameter